File: process.py

# The Process class represents a process in our system.

# When we print processes, we will print them in json format
import json

class Process:

    # Contructor - the default values for the member variables
    # are as indicated - an empty string for the process name,
    # and -1 for the process id
    #
    # TODO: 
    # Add the new attributes. Always initialize cpu_used to 0
    def __init__(self, process_name="", process_id=-1):
        self.process_name = process_name
        self.process_id = process_id

    
    # Set the process name 
    def setProcessName(self, process_name):
        self.process_name = process_name

    # Set the process ID
    def setProcessId(self, process_id):
        self.process_id = process_id

    # get the process name
    def processName(self):
        return self.process_name
    
    # get the process id
    def processId(self):
        return self.process_id


    # TODO:
    # Add setters and getters for the new
    # attributes (priority and cpu_needed)


    # TODO:
    # Add a getter (but not a setter) for cpu_used
    


    def run(self, num_cycles):
        # TODO:
        # Impmlement the run() method.
        #
        # Determine how many cycles the process should actually execute
        # (that number should either be the value of numCycles provided,
        # or a smaller number if the process doesn't need that many cycles)
        #
        # Update cpu_used to reflect the actual number of cycles executed, 
        # and return the number executed.



    # This method is called when print() is called. 
    # It converts the contents of the object
    # to json, and then dumps the json format. 
    #
    # TODO: 
    # Add the new attributes, including cpu_used
    def __str__(self):
        return json.dumps({"process_name" : self.process_name,
                           "process_id" : self.process_id, 
                           separators = (",",":"), sort_keys = True)

File: process.cpp
// process.cpp
//
// Include your implementation of the Process class here.

#include <string>
#include <iostream>
#include <sstream>

#include "process.h"
#include "nlohmann/json.hpp"

using namespace std;

// TODO: 
// Add the additional arguments so the new attributes can be initialized here.
// Initialize the attribute for cpu_used to 0
Process::Process(const string& processName, int priority) :
    m_processName(processName), 
    m_priority(priority)
{
    // Nothing to do here
}

const string& Process::processName() const
{
    return m_processName;
}

void Process::setProcessName(const string& processName)
{
    m_processName = processName;
}

const int Process::processId() const
{
    return m_processId;
}

void Process::setProcessId(int processId)
{
    m_processId = processId;
}


// TODO:
// Provide the implementations for the new setter/getter
// methods declared in the header file.



int Process::run(int numCycles)
{
    // TODO:
    // Impmlement the run() method.
    //
    // Determine how many cycles the process should actually execute
    // (that number should either be the value of numCycles provided,
    // or a smaller number if the process doesn't need that many cycles)
    //
    // Update cpu_used to reflect the actual number of cycles executed, 
    // and return the number executed.
}



// Friend function to print the object
std::ostream& operator<<(std::ostream& os, const Process& p)
{
    nlohmann::json j;
    to_json(j, p);
    os << j;
    return os;
}

// Functions to convert a process to and from json
void to_json(nlohmann::json& j, const Process& p)
{
    j["process_name"] = p.processName();
    j["process_id"] = p.processId();

    // TODO:
    // Add the new Process attributes to the output json

}

void from_json(const nlohmann::json& j, Process& p)
{
    if (j.contains("process_name")) {
        p.setProcessName(j["process_name"]);
    } else {
        ostringstream o;
        o << "Missing process_name in json input: " << j;
        throw(std::invalid_argument(o.str()));
    }

    if (j.contains("process_id")) {
        p.setProcessId(j["process_id"]);
    }

    // TODO: 
    // Retrieve the priority and cpu_needed attributes
    // from the input json and set them in the output
    // Process object
}